### ISSUE TITLE
LogicalType Unions

### ISSUE THREAD

**Issue by MichaelHirn** [mtth/avsc#297]
I ran into an unexpected `avsc` behavior - I guess the main question is, is this a bug or did I just use it incorrectly. At the bottom of the issue (*Question/1*) I added an excerpt from the docs that leaves the option open that I have just use it incorrectly.

## Situation

Given the following avdl file

```avdl
Protocol Test {

  @logicalType("LT")
  record A {
    string a;
  }

  @logicalType("LT")
  record B {
    string b;
  }

  @logicalType("LT")
  record C {
    union { A, B } c;
  }
}
```

and the following code

```javascript
// avro (aka. avsc) version: 5.4.19
avro.assembleProtocol(pathToAvdlFile, { delimitedCollections: true }, (err, schema) => {
  if (err) return reject(err);
  // we are creating a Service because
  // this allows us to get actual Type instances
  // via service.types
  const service = avro.Service.forProtocol(schema, { logicalTypes });
  // this code below is never reached because `avro.Service.forProtocol` fails
  schema.types = service.types;
  resolve(schema);
});
```

we get the following error

```
/.../avsc/lib/types.js:1192
      throw new Error(f('duplicate union branch name: %j', branch));
      ^

Error: duplicate union branch name: "record"
    at UnwrappedUnionType.<anonymous> (/.../avsc/lib/types.js:1192:13)
    at Array.forEach (<anonymous>)
    at UnwrappedUnionType.UnionType (/.../avsc/lib/types.js:1186:14)
    at new UnwrappedUnionType (/.../avsc/lib/types.js:1243:13)
    at Function.Type.forSchema (/.../avsc/lib/types.js:219:12)
    at new Field (/.../avsc/lib/types.js:2782:20)
    at RecordType.<anonymous> (/.../avsc/lib/types.js:2109:17)
    at Array.map (<anonymous>)
    at new RecordType (/.../avsc/lib/types.js:2108:45)
    at /.../avsc/lib/types.js:226:14
```

which - if we go to `types.js:1192` - is because the `branchName` of a `LogicalType` seems always to be `"record"`.

```javascript
var branch = type.branchName;
if (this._branchIndices[branch] !== undefined) {
  throw new Error(f('duplicate union branch name: %j', branch));
}
```

so if we naively check if `type` is a `LogicalType` and use its underlyingType it works.

```javascript
var branch = type.branchName;
if (type.underlyingType) { // there is probably a better way to check for LogicalType
  branch = type.underlyingType.branchName
}
if (this._branchIndices[branch] !== undefined) {
  throw new Error(f('duplicate union branch name: %j', branch));
}
```

Considering *Question/1* is this a bug or am I just using it wrong?

### Question/1

The API docs for `UnwrappedUnionType` notes:

> Finally, note that by using logical types, it is possible to work around the above requirements (by delegating the branch inference to the logical types themselves).

I couldn't find anything in the docs that mentioned how to specify branch inference in the `LogicalType` - curious if this relates to the issue and might be one way to solve this.

**Comment by mtth** [comment]
The error was a due to bug in logical type branch name derivation; it should be fixed in `5.4.21` (just released). Thanks for reporting it.

W.r.t. `UnwrappedUnionType` branch inference, this affects how unions are represented in-memory. Without logical types, unions with more than one record branch are always represented wrapped. With the schema above, this means valid in-memory instances of `C` are:

+ `{c: {A: {a: 'foo'}}`
+ `{c: {B: {b: 'bar'}}`

The wrapping object, which includes the chosen branch's name (`A` and `B` above), is used to efficiently keep track of which branch is present.

`avsc` provides more flexibility when logical types are used. In this case, inference is done by calling `isValid` on each logical type branch (which under the hood calls `_toValue`): if exactly one returns true, that branch is selected, otherwise an error is raised. This allows the more convenient unwrapped in-memory representations; here `C` instances would look like:

+ `{c: {a: 'foo'}}`
+ `{c: {b: 'bar'}}`

You can control which representation is used via the `wrapUnions` option (of `Type.forSchema` and `Service.forProtocol`). The default is `'auto'` which uses unwrapped when possible, falling back to wrapping if not.

**Comment by MichaelHirn** [comment]
Thanks for the quick turn around on the fix. `5.4.21` fixes the bug and will therefore close this issue.

Also appreciated the explanation of wrapped and unwrapped cases in correspondence with `LogicalType` - makes total sense now.
